### PR TITLE
Early stopping doc

### PR DESCRIPTION
**Pull request by Lucas-Steinmann**
## What does this PR do?

Improve documentation for early stopping with Result classes.
Remove some out-of-date or wrong documentation relating early stopping.

Fixes (no issue, only doc improvements)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements) No. It was only discussed on Slack.
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/results.rst [hunk 1]
```diff
@@ -97,17 +97,6 @@ checkpointing or early stopping:
         return TrainResult(some_metric, checkpoint_on=metric_a, early_stop_on=metric_b)
 
 
-In the manual loop, checkpoint and early stop is based only on the loss returned. With the `TrainResult` you
```
Comment (williamFalcon) [inline docs/source/results.rst:100]: why is this being removed? this whole block is still true

Comment (Lucas-Steinmann) [reply docs/source/results.rst:100]: Now I'm not sure, if I understood the block correctly. So maybe you can quickly answer me:

Why can I only use the "loss"?
The test here even says: https://github.com/PyTorchLightning/pytorch-lightning/blob/bd35c869ee0b856fc30229ef36957b058fe97461/tests/callbacks/test_early_stopping.py#L141-L143.

And why can't I change that every batch?
Maybe I somehow completely miss the point of the topic. If so I hapilly undo this change.

Comment (Lucas-Steinmann) [reply docs/source/results.rst:100]: Ok. I revise my argumentation.
I take everything back and claim the opposite.
I think early stopping only using train return dicts doesn't work at all. 
Not even with 'loss'.
The test case I mentioned above has a bug.

It checks whether current_epoch is smaller than max_epoch (presumably to check if early stopping was triggered).
But this check will always be true, since epochs are counted from 0.
I'll make a new issue, since now we are changing code and I guess it also requires discussion which cases should be supported (any keywords as suggested in the test, only 'loss' as by the docs and you or keep it as it is)

Comment (Lucas-Steinmann) [reply docs/source/results.rst:100]: Here is a minimal example, to show you that it doesn't work with "loss":

```python3
import pytorch_lightning as pl
import torch
from torch import Tensor
from torch.optim import Optimizer
from torch.utils.data import DataLoader


class MyModule(pl.LightningModule):

    def training_step(self, batch, batch_idx):
        losses = [8, 4, 2, 2, 4, 5, 6, 8, 10, 11]
        return {'loss': torch.tensor(float(losses[self.current_epoch]))}

    def configure_optimizers(self):
        pass

    def backward(self, trainer, loss: Tensor, optimizer: Optimizer, optimizer_idx: int) -> None:
        pass


model = MyModule()
early_stopping = pl.callbacks.EarlyStopping(monitor='loss', patience=0)
trainer = pl.Trainer(early_stop_callback=early_stopping, max_epochs=10)
trainer.fit(model, DataLoader(list(range(100))))
print(trainer.current_epoch)
```

This will be current epoch as 9, but should be lower. See (and please reopen) #3199

File: docs/source/results.rst [hunk 2]
```diff
@@ -80,10 +80,6 @@ The `TrainResult` basic usage is this:
 
 minimize
 ^^^^^^^^
-When using TrainResult, the metric that needs to be minimized is passed to this
```
Comment (Borda) [inline docs/source/results.rst:83]: why this doc is removed?

Comment (Lucas-Steinmann) [reply docs/source/results.rst:83]: Sorry. I think I oversaw this change, when doing the rebase. I will fix this.

File: docs/source/results.rst (hunk 1 above)
Comment (Lucas-Steinmann) [reply docs/source/results.rst:104]: I will remove this change and we can fix the docs in a new pull request, where we fix issue #3199 
That way, we can get through with this pull request faster.

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/trainer/__init__.py [hunk 3]
```diff
@@ -411,10 +411,13 @@ def on_train_end(self, trainer, pl_module):
 Callback for early stopping.
 early_stop_callback (:class:`pytorch_lightning.callbacks.EarlyStopping`)
 
-- ``True``: A default callback monitoring ``'val_loss'`` is created.
-   Will raise an error if ``'val_loss'`` is not found.
+- ``True``: A default callback monitoring ``'val_loss'`` (if dict is returned in validation loop) or
+   ``early_stopping_on`` (if :class:`~pytorch_lightning.core.step_result.Result` is returned) is created.
+   Will raise an error if a dictionary is returned and ``'val_loss'`` is not found.
+   Will raise an error if a :class:`~pytorch_lightning.core.step_result.Result` is returned
+   and ``early_stopping_on`` was not specified.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/__init__.py:418]:
```suggestion
- ``True``: A default callback monitoring ``'val_loss'`` (if dict is returned in validation loop) or
  ``early_stopping_on`` (if :class:`~pytorch_lightning.core.step_result.Result` is returned) is created.
  Will raise an error if a dictionary is returned and ``'val_loss'`` is not found.
  Will raise an error if a :class:`~pytorch_lightning.core.step_result.Result` is returned
  and ``early_stopping_on`` was not specified.
```

File: pytorch_lightning/core/step_result.py [hunk 4]
```diff
@@ -499,7 +499,16 @@ def training_step(self, batch, batch_idx):
         Args:
             minimize: Metric currently being minimized.
             early_stop_on: Metric to early stop on.
+                Should be singleton tensor if combined with default
```
[the remark this answers is not in the document]
Comment (Lucas-Steinmann) [reply pytorch_lightning/core/step_result.py]: You are correct. Yes, I replaced it with the terminology used in the error message of pytorch: "one element tensor".

File: docs/source/early_stopping.rst [hunk 5]
```diff
@@ -33,9 +33,17 @@ callback can be used to monitor a validation metric and stop the training when n
...
-    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`
+    If a dict is returned by
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`,
+    the callback will look for `val_loss` in the dict
     and raise an error if `val_loss` is not present.
+    Otherwise, if a :class:`~pytorch_lightning.core.step_result.Result` is returned, by
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`,
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step` or
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`,
+    the `early_stop_on` metric, specified in the initialization of the
```
[the remark this answers is not in the document]
Comment (Lucas-Steinmann) [reply docs/source/early_stopping.rst:44]: In that case early stopping is not triggered for me. I wanted to ask, if that is the expected behaviour or not after my other issue related to early stopping (#3199) is fixed. I didn't want to swamp you with issues :). Here is an example:

https://gist.github.com/Lucas-Steinmann/f900b3ef5636028d5917d4cb0183a291

File: docs/source/early_stopping.rst [hunk 6]
```diff
@@ -33,9 +33,17 @@ callback can be used to monitor a validation metric and stop the training when n
...
+    :class:`~pytorch_lightning.core.step_result.Result` object is used
+    and raise an error if it was not specified.
```
[the remark this answers is not in the document]
Comment (Lucas-Steinmann) [reply docs/source/early_stopping.rst]: You are correct. Strict checking is disabled with structured results. I've replaced it with ``and display a warning if it was not specified.``. I'll try to be more careful :)

File: docs/source/early_stopping.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/early_stopping.rst:44]: ok good to know, let's fix this separately and later update the docs here

File: pytorch_lightning/core/step_result.py [hunk 7]
```diff
@@ -499,7 +499,16 @@ def training_step(self, batch, batch_idx):
         Args:
             minimize: Metric currently being minimized.
             early_stop_on: Metric to early stop on.
+                Should be a one element tensor if combined with default
+                :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`.
+                If this result is returned by
+                :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`,
+                the specified value will be averaged across all steps.
             checkpoint_on: Metric to checkpoint on.
+                Should be a one elemen tensor if combined with default checkpoint callback.
```
Comment (awaelchli) [inline pytorch_lightning/core/step_result.py:508]:
```suggestion
                Should be a one element tensor if combined with default checkpoint callback.
```

File: pytorch_lightning/core/step_result.py [hunk 8]
```diff
@@ -654,7 +663,16 @@ def test_step(self, batch, batch_idx):
 
         Args:
             early_stop_on: Metric to early stop on.
+                Should be singleton tensor if combined with default
```
Comment (awaelchli) [inline pytorch_lightning/core/step_result.py:666]:
```suggestion
                Should be a one element tensor if combined with default
```

File: pytorch_lightning/core/step_result.py [hunk 9]
```diff
@@ -654,7 +663,16 @@ def test_step(self, batch, batch_idx):
...
+                :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`.
+                If this result is returned by
+                :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`,
+                the specified value will be averaged across all steps.
             checkpoint_on: Metric to checkpoint on.
+                Should be singleton tensor if combined with default checkpoint callback.
```
Comment (awaelchli) [inline pytorch_lightning/core/step_result.py:672]:
```suggestion
                Should be a one element tensor if combined with default checkpoint callback.
```

File: docs/source/early_stopping.rst [hunk 10]
```diff
@@ -33,9 +33,17 @@ callback can be used to monitor a validation metric and stop the training when n
 There are two ways to enable the EarlyStopping callback:
 
 -   Set `early_stop_callback=True`.
-    The callback will look for 'val_loss' in the dict returned by 
-    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`
+    If a dict is returned by
+    :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`,
+    the callback will look for `val_loss` in the dict
     and raise an error if `val_loss` is not present.
```
Comment (awaelchli) [inline docs/source/early_stopping.rst:39]:
```suggestion
    and display a warning if `val_loss` is not present.
```

File: docs/source/early_stopping.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lucas-Steinmann) [reply docs/source/early_stopping.rst:44]: I have started doing so with issue #3413 and PR #3414

**Review by Borda — APPROVED** [review]
LGTM 🦝
